### PR TITLE
Remove Mocha and use rspec-mocks instead

### PR DESCRIPTION
**Pull request by jhelwig**
The version of Mocha we're using is ancient, and documentation for this version is nearly impossible to find. Since the last time the version of Mocha we're using has been updated, RSpec has gained the ability to do all of the mocking & stubbing that Mocha provided, so there isn't a compelling reason to use Mocha beyond "that's what we've always used". Since we've been moving towards recommending using rspec-mocks for testing things like modules, it'd probably be a good idea to be consistent across projects and use rspec-mocks in Puppet as well. Plus, it means that we'll have one fewer dependency for testing Puppet.

### CODE REVIEW

**Comment by pcarlisle** [comment]
Looks like it's impossible to respond on changes directly due to the size of this.

I can only really do a rough sample searching for things, but as far as I can tell all the stuff I remember looking problematic has been taken care of. The biggest thing I remember needing more manual work was cases where both with and returns used a block and that looks like it's handled.

Minor:
You rewrote `responds_with` using blocks, you can use `having_attributes` and convert arguments to hash form
eg
```ruby
responds_with(:key, 'puppet.node.test')
```
becomes
```ruby
having_attributes(key: 'puppet.node.test')
```

**Comment by pcarlisle** [comment]
better to merge this as soon as it passes though and then refine after

File: docs/rspec_tutorial.md [hunk 1]
```diff
@@ -40,11 +40,11 @@ testing documentation](acceptance_tests.md)
...
 
-Puppet is only compatible with certain versions of RSpec and Mocha. If you are
-not using Bundler to install the required test libraries you must ensure that
-you are using the right library versions. Using unsupported versions of Mocha
-and RSpec will probably display many spurious failures. The supported versions
-of RSpec and Mocha can be found in the project Gemfile.
+Puppet is only compatible with a specific version of RSpec. If you are not
+using Bundler to install the required test libraries you must ensure that you
+are using the right library version. Using an unsupported version of RSpec will
+probably display many spurious failures. The supported version of RSpec can be
```
Comment (melissa) [inline docs/rspec_tutorial.md]: `misleading` is a little more approachable than `spurious`

Comment (melissa) [reply docs/rspec_tutorial.md]: gonna address this later

**Comment by joshcooper** [comment]
Not blocking at all, but after merging this and up, there are a few shebang stragglers in 5.5.x and later branches

```
spec/unit/status_spec.rb:#! /usr/bin/env ruby
```

**Comment by andyw8** [comment]
@jhelwig I'm curious to hear what tooling/approach you used for this large conversion.

**Pull request merged**